src/requestManager.cpp

```diff
@@ -984,65 +1021,108 @@ void CRequestManager::FindNextBlocksToDownload(NodeId nodeid, unsigned int count
+    if (itInFlight == mapBlocksInFlight.end() || !mapBlocksInFlight[hash].count(nodeid))
     {
+        // Get a request manager nodestate pointer.
+        std::map<NodeId, CRequestManagerNodeState>::iterator it = mapRequestManagerNodeState.find(nodeid);
+        DbgAssert(it != mapRequestManagerNodeState.end(), return );
+        CRequestManagerNodeState *state = &it->second;
+
+        // Add queued block to nodestate and add iterator for queued block to mapBlocksInFlight
         int64_t nNow = GetTimeMicros();
         QueuedBlock newentry = {hash, nNow};
-        std::list<QueuedBlock>::iterator it = state->vBlocksInFlight.insert(state->vBlocksInFlight.end(), newentry);
+        std::list<QueuedBlock>::iterator it2 = state->vBlocksInFlight.insert(state->vBlocksInFlight.end(), newentry);
+        mapBlocksInFlight[hash][nodeid] = it2;
+
+        // Increment blocks in flight for this node and if applicable the time we started downloading.
         state->nBlocksInFlight++;
         if (state->nBlocksInFlight == 1)
         {
             // We're starting a block download (batch) from this peer.
             state->nDownloadingSince = GetTimeMicros();
         }
-        mapBlocksInFlight[hash] = std::make_pair(nodeid, it);
     }
 }
 
 // Returns a bool if successful in indicating we received this block.
 bool CRequestManager::MarkBlockAsReceived(const uint256 &hash, CNode *pnode)
 {
-    AssertLockHeld(cs_main);
+    if (!pnode)
+        return false;
 
     LOCK(cs_objDownloader);
-    std::map<uint256, std::pair<NodeId, std::list<QueuedBlock>::iterator> >::iterator itInFlight =
+    NodeId nodeid = pnode->GetId();
+    std::map<uint256, std::map<NodeId, std::list<QueuedBlock>::iterator> >::iterator itInFlight =
         mapBlocksInFlight.find(hash);
-    if (itInFlight != mapBlocksInFlight.end())
+    if (itInFlight != mapBlocksInFlight.end() && mapBlocksInFlight[hash].count(nodeid))
     {
-        CNodeState *state = State(itInFlight->second.first);
-        DbgAssert(state != nullptr, return false);
+        // Get a request manager nodestate pointer.
+        std::map<NodeId, CRequestManagerNodeState>::iterator it = mapRequestManagerNodeState.find(nodeid);
+        DbgAssert(it != mapRequestManagerNodeState.end(), return false);
+        CRequestManagerNodeState *state = &it->second;
 
-        int64_t getdataTime = itInFlight->second.second->nTime;
+        int64_t getdataTime = mapBlocksInFlight[hash][nodeid]->nTime;
         int64_t now = GetTimeMicros();
         double nResponseTime = (double)(now - getdataTime) / 1000000.0;
 
         // calculate avg block response time over a range of blocks to be used for IBD tuning.
-        static uint8_t blockRange = 50;
+        uint8_t blockRange = 50;
         {
             LOCK(pnode->cs_nAvgBlkResponseTime);
             if (pnode->nAvgBlkResponseTime < 0)
-                pnode->nAvgBlkResponseTime = 2.0;
+                pnode->nAvgBlkResponseTime = 0.0;
```

Comment:
this had to change otherwise we'd be disconnecting pretty much every peer during the first part of initial sync since their average response time would start out quite high (2 seconds) vs the average which is typically just .5 or less.  Now that we're disconnecting the slow peers it's not an issue that we ask for 64 blocks from them on the first try and it takes more than 50 anyway before we check that peer and issue a disconnect.

Reply:
fair enough